packages/test/test-end-to-end-tests/src/test/dataMigration.spec.ts

```diff
@@ -0,0 +1,425 @@
+	public get containerRuntime() {
+		return this.context.containerRuntime as ContainerRuntime;
+	}
+	public get getter(): SharedCell {
+		assert(this._getter !== undefined, "unexpected get of getter before it was defined");
+		return this._getter;
+	}
+	private _getter?: SharedCell;
+
+	public get rootDDS() {
+		return this.root;
+	}
+
+	protected async initializingFirstTime(props?: any): Promise<void> {
+		const getter = SharedCell.create(this.runtime);
+		this.root.set(getterKey, getter.handle);
+		this._getter = getter;
+		this._getter.set("abc");
+	}
+
+	protected async initializingFromExisting(): Promise<void> {
+		const handle = this.root.get<IFluidHandle<SharedCell>>(getterKey);
+		this._getter = await handle?.get();
+	}
+}
+
+class DataObjectV2 extends DataObject {
+	public get containerRuntime() {
+		return this.context.containerRuntime as ContainerRuntime;
+	}
+	public get getter(): SharedMap {
+		assert(this._getter !== undefined, "unexpected get of getter before it was defined");
+		return this._getter;
+	}
+	private _getter?: SharedMap;
+	public get rootDDS() {
+		return this.root;
+	}
+
+	protected async initializingFirstTime(props?: any): Promise<void> {
+		const getter = SharedMap.create(this.runtime);
+		this.root.set(getterKey, getter.handle);
+		this._getter = getter;
+	}
+
+	protected async initializingFromExisting(): Promise<void> {
+		const handle = this.root.get<IFluidHandle<SharedMap>>(getterKey);
+		this._getter = await handle?.get();
+	}
+}
+
+function createReadOnlyContext(
+	context: IContainerContext,
+	codeDetailsVersion: string,
+): IContainerContext {
+	const readonlyContext: IContainerContext = {
+		...context,
+		connected: true,
+		clientDetails: {
+			capabilities: context.clientDetails.capabilities,
+			type: detachedClientType,
+		},
+		storage: {
+			...context.storage,
+			getSnapshotTree: context.storage.getSnapshotTree,
+			getVersions: context.storage.getVersions,
+			createBlob: async (_file: ArrayBufferLike) => {
+				throw new Error("Should not be submitting blobs to storage");
+			},
+			readBlob: context.storage.readBlob,
+			uploadSummaryWithContext: async (_summary, _context) => {
+				throw new Error("Should not be uploading a summary");
+			},
+			downloadSummary: context.storage.downloadSummary,
+		},
+		submitFn: (_message, _contents, _batch, _appData) => 0,
+		submitSummaryFn: (_summaryOp, _referenceSequenceNumber?) => 0,
+		submitBatchFn: (_batch, _referenceSequenceNumber?) => 0,
+		submitSignalFn: (_contents: any) => {},
+		disposeFn: context.disposeFn,
+		closeFn: context.closeFn,
+		updateDirtyContainerState: context.updateDirtyContainerState,
+		getAbsoluteUrl: context.getAbsoluteUrl,
+		getSpecifiedCodeDetails: () => {
+			return { package: codeDetailsVersion };
+		},
+		getLoadedFromVersion: context.getLoadedFromVersion,
+		attachState: context.attachState,
+	};
+	return readonlyContext;
+}
+
+class ContainerRuntimeFactoryWithDataMigration extends RuntimeFactoryHelper {
```

Comment:
might be worth adding comments above key classes to clarify their purpose.
If I got it right, this is an CRF adapter, that makes a call on what CRF instance to use depending on version and detached/attached state.